### PR TITLE
drakrun: lvm backend unit tests

### PR DESCRIPTION
**Pull request by manorit2001**
This PR references #473 and includes LVM unit tests along with bug rectifications wherever found

---
This change is [Reviewable](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480)

### CODE REVIEW

**Comment by chivay** [comment]
___
*[.drone.yml, line 39 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVptoB07-FqpNj293pg:-MVptoB07-FqpNj293ph:b-ac8bc2) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/.drone.yml#L39)):*
> ```YAML
>   commands:
>     - apt-get update && apt-get install -y python3-parted libparted-dev
>     - pip3 install pytest pyparted
> ```

At this point we can create `requirements.txt` in `test/` and just install the dependencies for required testing.

**Comment by chivay** [comment]
BTW, I'm testing reviewable as it seems superior to GH reviews

**Comment by manorit2001** [comment]
> _[.drone.yml, line 39 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVptoB07-FqpNj293pg:-MVptoB07-FqpNj293ph:b-ac8bc2) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/.drone.yml#L39)):_
> 
> > ```yaml
> >   commands:
> >     - apt-get update && apt-get install -y python3-parted libparted-dev
> >     - pip3 install pytest pyparted
> > ```
> 
> At this point we can create `requirements.txt` in `test/` and just install the dependencies for required testing.

Seems right, will do it

**Comment by chivay** [comment]
___
*[drakrun/drakrun/test/test_lvm.py, line 137 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVq-YO795lIqzT4beck:-MVq-YO795lIqzT4becl:b77xime) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/test/test_lvm.py#L137)):*
> ```Python
> 
> @pytest.fixture
> def create_partitions(mount_vm0):
> ```

This code looks quite nice, but after we merge your PR that uses Injector instead of mounting, quite a lot of things will be simplified.
Basically the `mount_vm0_root_as_block` will become dead code and I think that this method could be removed and thus testing will be unnecessary.
What do you think? :thinking:

**Comment by chivay** [comment]
___
*[drakrun/drakrun/test/test_lvm.py, line 92 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVpzbSuEZWlFessOU86:-MVpzbSuEZWlFessOU87:b-mie9kk) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/test/test_lvm.py#L92)):*
> ```Python
>     # Should not raise any exceptions
>     logging.info("Get snapshot time")
>     backend.get_vm0_snapshot_time()
> ```

We should do at least basic sanity check such as the return value type

**Comment by chivay** [comment]
___
*[drakrun/drakrun/test/test_lvm.py, line 56 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVpyE9g3m_CyxhTQlef:-MVpyE9g3m_CyxhTQleg:bfii16p) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/test/test_lvm.py#L56)):*
> ```
>     backup(install_info)
>     yield LvmStorageBackend(install_info)
>     restore(install_info)
> ```

I agree with the comment you've added above - `# best way would be to use a new lvm volume group for testing`
Test code shouldn't assume that the sandbox is properly configured. We should fake the `InstallInfo` and setup a new volume group just for testing

**Comment by chivay** [comment]
___
*[drakrun/drakrun/storage.py, line 543 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVpwzV624J79zfSJguT:-MVpwzV624J79zfSJguU:b-norc87) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/storage.py#L543)):*
> ```Python
>         )
> 
>     def delete_vm_volume(self, vm_id: str):
> ```

Please, don't mix different changes into the same PR.
It introduces unnecessary noise.

**Comment by manorit2001** [comment]
> _[drakrun/drakrun/test/test_lvm.py, line 137 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVq-YO795lIqzT4beck:-MVq-YO795lIqzT4becl:b77xime) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/test/test_lvm.py#L137)):_
> 
> > ```python
> > @pytest.fixture
> > def create_partitions(mount_vm0):
> > ```
> 
> This code looks quite nice, but after we merge your PR that uses Injector instead of mounting, quite a lot of things will be simplified.
> Basically the `mount_vm0_root_as_block` will become dead code and I think that this method could be removed and thus testing will be unnecessary.
> What do you think?

Sure, I really don't have issues with that, As long as `mount_vm0_root_as_block` also gets removed. I can remove that block from my injector PR if that isn't necessary anymore

**Comment by manorit2001** [comment]
> _[drakrun/drakrun/test/test_lvm.py, line 92 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVpzbSuEZWlFessOU86:-MVpzbSuEZWlFessOU87:b-mie9kk) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/test/test_lvm.py#L92)):_
> 
> > ```python
> >     # Should not raise any exceptions
> >     logging.info("Get snapshot time")
> >     backend.get_vm0_snapshot_time()
> > ```
> 
> We should do at least basic sanity check such as the return value type

Sure, sounds nice, will do it

**Comment by manorit2001** [comment]
> _[drakrun/drakrun/storage.py, line 543 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVpwzV624J79zfSJguT:-MVpwzV624J79zfSJguU:b-norc87) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/storage.py#L543)):_
> 
> > ```python
> >         )
> > 
> >     def delete_vm_volume(self, vm_id: str):
> > ```
> 
> Please, don't mix different changes into the same PR.
> It introduces unnecessary noise.

It was actually necessary for testing sir, coz the delete block also needed to be tested and unit test for that also needed to be there. That's why included it here

**Review by manorit2001 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480)* status: 0 of 4 files reviewed, 5 unresolved discussions (waiting on @chivay)

___
*[drakrun/drakrun/test/test_lvm.py, line 56 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVpyE9g3m_CyxhTQlef:-MVrPwlI6vxQIrYySYAU:bmbm8aj) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/test/test_lvm.py#L56)):*
<details><summary><i>Previously, chivay (Hubert Jasudowicz) wrote…</i></summary><blockquote>

> ```
>     backup(install_info)
>     yield LvmStorageBackend(install_info)
>     restore(install_info)
> ```

I agree with the comment you've added above - `# best way would be to use a new lvm volume group for testing`
Test code shouldn't assume that the sandbox is properly configured. We should fake the `InstallInfo` and setup a new volume group just for testing

</blockquote></details>

I was actually working on patching this but I am having issues actually with that, the main issue is that we can't just make a volume group, It can become a very complicated code(using basic lvm commands) if we want to setup a lvm volume group by ourselves. The other solution can be that during the test, we can use some env variable or pass a argument with a created volume group which will be used in the test. This might be easier. What do you think on these?

**Review by chivay — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480)* status: all files reviewed, 4 unresolved discussions (waiting on @manorit2001)

___
*[.drone.yml, line 39 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVptoB07-FqpNj293pg:-MVuw5-s8RFxRLJfI8K7:b-8fx2u7) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/.drone.yml#L39)):*
<details><summary><i>Previously, manorit2001 (Manorit Chawdhry) wrote…</i></summary><blockquote>

Seems right, will do it
</blockquote></details>

:+1: 
___
*[drakrun/drakrun/storage.py, line 543 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVpwzV624J79zfSJguT:-MVuw1zjBkoOaCUYs0U1:b32w0x) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/storage.py#L543)):*
<details><summary><i>Previously, manorit2001 (Manorit Chawdhry) wrote…</i></summary><blockquote>

It was actually necessary for testing sir, coz the delete block also needed to be tested and unit test for that also needed to be there. That's why included it here
</blockquote></details>

OK
___
*[drakrun/drakrun/test/test_lvm.py, line 56 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVpyE9g3m_CyxhTQlef:-MVuvFxs0l0I9Vqil9yo:bpcvjaw) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/test/test_lvm.py#L56)):*
<details><summary><i>Previously, manorit2001 (Manorit Chawdhry) wrote…</i></summary><blockquote>

I was actually working on patching this but I am having issues actually with that, the main issue is that we can't just make a volume group, It can become a very complicated code(using basic lvm commands) if we want to setup a lvm volume group by ourselves. The other solution can be that during the test, we can use some env variable or pass a argument with a created volume group which will be used in the test. This might be easier. What do you think on these?
</blockquote></details>

On the surface it may seem complex but we don't have to use physical block device to access LVM features. We just have to perform a few simple steps:
1. Create a temporary file with some sensible size, e.g. 50M
2. Call losetup and bind it to /dev/loop* -> now we have a block device that is mapped to a file
3.  At this point we can run vgcreate on /dev/loop* and there we have it

In this way we don't depend on an existing installation and setup the environment ourselves
___
*[drakrun/drakrun/test/test_lvm.py, line 92 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVpzbSuEZWlFessOU86:-MVuvDbIDlariXF-jvA4:b-8fx2u7) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/test/test_lvm.py#L92)):*
<details><summary><i>Previously, manorit2001 (Manorit Chawdhry) wrote…</i></summary><blockquote>

Sure, sounds nice, will do it
</blockquote></details>

:+1: 
___
*[drakrun/drakrun/test/test_lvm.py, line 137 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVq-YO795lIqzT4beck:-MVuvBBs8uQ0L5F1r9sd:b-qhjy86) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/test/test_lvm.py#L137)):*
<details><summary><i>Previously, manorit2001 (Manorit Chawdhry) wrote…</i></summary><blockquote>

Sure, I really don't have issues with that, As long as `mount_vm0_root_as_block` also gets removed. I can remove that block from my injector PR if that isn't necessary anymore
</blockquote></details>

Sounds good :+1:

**Review by manorit2001 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480)* status: all files reviewed, 4 unresolved discussions (waiting on @chivay and @manorit2001)

___
*[.drone.yml, line 39 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVptoB07-FqpNj293pg:-MVv71AZ-EYrA85KZZa2:b-e7tm33) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/.drone.yml#L39)):*
<details><summary><i>Previously, chivay (Hubert Jasudowicz) wrote…</i></summary><blockquote>

:+1: 
</blockquote></details>

Will this be necessary anymore btw? Coz these changes will have to be reverted only as the unit tests which create the partitions is no longer required
___
*[drakrun/drakrun/test/test_lvm.py, line 56 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVpyE9g3m_CyxhTQlef:-MVv6X0L4MdnQ48pMynj:beach56) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/test/test_lvm.py#L56)):*
<details><summary><i>Previously, chivay (Hubert Jasudowicz) wrote…</i></summary><blockquote>

On the surface it may seem complex but we don't have to use physical block device to access LVM features. We just have to perform a few simple steps:
1. Create a temporary file with some sensible size, e.g. 50M
2. Call losetup and bind it to /dev/loop* -> now we have a block device that is mapped to a file
3.  At this point we can run vgcreate on /dev/loop* and there we have it

In this way we don't depend on an existing installation and setup the environment ourselves
</blockquote></details>

Woah, didn't think about loopback files. This sounds doable now. Will do it, thanks!

**Review by manorit2001 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480)* status: 3 of 4 files reviewed, 3 unresolved discussions (waiting on @chivay)

___
*[drakrun/drakrun/test/test_lvm.py, line 56 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVpyE9g3m_CyxhTQlef:-MVvPl0Q0H6IADoN60NW:b-896fix) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/test/test_lvm.py#L56)):*
<details><summary><i>Previously, manorit2001 (Manorit Chawdhry) wrote…</i></summary><blockquote>

Woah, didn't think about loopback files. This sounds doable now. Will do it, thanks!
</blockquote></details>

Done.
___
*[drakrun/drakrun/test/test_lvm.py, line 92 at r1](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480#-MVpzbSuEZWlFessOU86:-MVvPzwFBMZCIFbWEjqd:b-896fix) ([raw file](https://github.com/cert-polska/drakvuf-sandbox/blob/5bcd4fe888fdae615c20a9fa152c5b376b34ca21/drakrun/drakrun/test/test_lvm.py#L92)):*
<details><summary><i>Previously, chivay (Hubert Jasudowicz) wrote…</i></summary><blockquote>

:+1: 
</blockquote></details>

Done.

**Comment by chivay** [comment]
> Ugh! the changes are across multiple PRs, it is difficult to work. I need the `common_utils` file now which is there in #481. Any idea on how to manage all these things?
> 
> @chivay @icedevml


You can pull it out into a separate branch, and rebase other changes on top of it

File: .drone.yml [hunk 1]
```diff
@@ -35,7 +35,8 @@ steps:
 - name: run drakrun tests
   image: python:3.8-buster
   commands:
-    - pip3 install pytest
+    - apt-get update && apt-get install -y python3-parted libparted-dev
+    - pip3 install pytest pyparted
```
Comment (chivay) [inline .drone.yml:39]: parted is now unnecessary

File: drakrun/drakrun/test/test_lvm.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    logging.basicConfig(
+        level=logging.DEBUG,
+        format='[%(asctime)s][%(levelname)s] %(message)s',
+        handlers=[logging.StreamHandler()]
+    )
+
+
+@pytest.fixture(scope="module")
+def backend(setup):
+    yield LvmStorageBackend(InstallInfo.load())
```
Comment (chivay) [inline drakrun/drakrun/test/test_lvm.py:64]: This fixture has no cleanup, it can return instead of yielding

File: drakrun/drakrun/test/test_lvm.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    safe_delete(temp_file_name)
+
+
+@pytest.fixture(autouse=True)
+def enable_logging():
+    logging.basicConfig(
+        level=logging.DEBUG,
+        format='[%(asctime)s][%(levelname)s] %(message)s',
+        handlers=[logging.StreamHandler()]
+    )
```
Comment (chivay) [inline drakrun/drakrun/test/test_lvm.py:59]: pytest handles logging automatically, doing it manually in a fixture may result in some weird behavior

File: drakrun/drakrun/test/test_lvm.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+@pytest.fixture(scope="module")
+def setup():
+    temp_file_name = tempfile.NamedTemporaryFile(delete=False).name
+    print(temp_file_name)
+    subprocess.run(['dd', 'if=/dev/zero', f'of={temp_file_name}', 'bs=1M', 'count=100'], stderr=subprocess.STDOUT, check=True)
+    loopback_file = subprocess.check_output(['losetup', '-f', '--show', temp_file_name], stderr=subprocess.STDOUT).decode().strip('\n')
+
+    # v is being added in start to ensure the lvm volume group starts with a letter
+    # don't know if it is required or not
+    volume_group = 'v' + ''.join(secrets.choice(string.ascii_letters + string.digits) for i in range(5))
```
Comment (chivay) [inline drakrun/drakrun/test/test_lvm.py:25]: I'd prefer the tests to be deterministic. Please choose some simple name for the volume group.

```
# v is being added in start to ensure the lvm volume group starts with a letter
# don't know if it is required or not
```

I don't understand this. What's the origin of "v" ?

File: drakrun/drakrun/test/test_lvm.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            storage_backend='lvm',
+            disk_size='25M',
+            iso_path=None,  # not being required
+            zfs_tank_name=None,
+            lvm_volume_group=volume_group,
+            enable_unattended=None,
+            iso_sha256=None
+        )
+    InstallInfo.load = install_patch
+    InstallInfo.try_load = install_patch
```
Comment (chivay) [inline drakrun/drakrun/test/test_lvm.py:43]: This will break InstallInfo class throughout the execution.
pytest has dedicated support for monkeypatching and mocking, please use it.
Additionaly, InstallInfo could be provided as a fixture to functions that load is explicitly

**Review by chivay — COMMENTED** [review]
I'm worried about implicit dependencies between tests, such as:
`test_snapshot_lvm` -> `test_time`
This should be done by fixtures.

File: drakrun/drakrun/test/test_lvm.py (hunk 4 above)
Comment (manorit2001) [reply drakrun/drakrun/test/test_lvm.py:25]: No origin

> I'd prefer the tests to be deterministic. Please choose some simple name for the volume group.
> 
> ```
> # v is being added in start to ensure the lvm volume group starts with a letter
> # don't know if it is required or not
> ```
> 
> I don't understand this. What's the origin of "v" ?

Just a random letter (Inspired by "v"olume group), no specific origin as such. What name do you suggest for this then?

File: drakrun/drakrun/test/test_lvm.py (hunk 2 above)
Comment (manorit2001) [reply drakrun/drakrun/test/test_lvm.py:64]: > I'm worried about implicit dependencies between tests, such as:
> `test_snapshot_lvm` -> `test_time`
> This should be done by fixtures.

Could you please explain it more?

**Comment by manorit2001** [comment]
> I'm worried about implicit dependencies between tests, such as:
> `test_snapshot_lvm` -> `test_time`
> This should be done by fixtures.

Could you please explain it more?

File: drakrun/drakrun/test/test_lvm.py (hunk 5 above)
Comment (manorit2001) [reply drakrun/drakrun/test/test_lvm.py:43]: We need the new `InstallInfo` only throughout the execution, we don't need the original one. So, isn't it safe to just override this?

File: drakrun/drakrun/test/test_lvm.py (hunk 2 above)
Comment (manorit2001) [reply drakrun/drakrun/test/test_lvm.py:64]: If this one returns, then will the `setup` cleanup run at the end? Coz we don't want the `setup` cleanup to be executed as soon as we return from this

**Comment by icedevml** [comment]
@chivay https://docs.pytest.org/en/stable/example/simple.html#incremental-testing-test-steps
Looks like this is some pattern in pytest called "incremental testing" (I was also not aware of it till today)

**Review by icedevml — APPROVED** [review]
> Reviewed 1 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cert-polska/drakvuf-sandbox/480)* status: 0 of 6 files reviewed, 8 unresolved discussions (waiting on @chivay and @manorit2001)

**Review by chivay — APPROVED** [review]
Looks great! Thank you for your hard work :)
